File: examples/bindings/python/openfiles_client.py
"""Tiny stdlib OpenFiles HTTP client."""
from __future__ import annotations

import json
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any


@dataclass
class OpenFilesClient:
    base_url: str = "http://127.0.0.1:8787"

    def _url(self, prefix: str, path: str = "") -> str:
        path = path.strip("/")
        if path:
            return f"{self.base_url}{prefix}/{urllib.parse.quote(path)}"
        return f"{self.base_url}{prefix}"

    def list(self, path: str = "/") -> list[dict[str, Any]]:
        with urllib.request.urlopen(self._url("/v1/list", path)) as resp:
            return json.loads(resp.read().decode())

    def stat(self, path: str) -> dict[str, Any]:
        with urllib.request.urlopen(self._url("/v1/stat", path)) as resp:
            return json.loads(resp.read().decode())

    def read(self, path: str, offset: int | None = None, length: int | None = None) -> bytes:
        url = self._url("/v1/read", path)
        if offset is not None and length is not None:
            url += f"?offset={offset}&len={length}"
        with urllib.request.urlopen(url) as resp:
            return resp.read()

    def write(self, path: str, data: bytes) -> None:
        req = urllib.request.Request(self._url("/v1/write", path), data=data, method="PUT")
        with urllib.request.urlopen(req) as resp:
            resp.read()

    def delete(self, path: str) -> None:
        req = urllib.request.Request(self._url("/v1/delete", path), method="DELETE")
        with urllib.request.urlopen(req) as resp:
            resp.read()

    def flush(self) -> int:
        req = urllib.request.Request(f"{self.base_url}/v1/flush", data=b"", method="POST")
        with urllib.request.urlopen(req) as resp:
            return int(json.loads(resp.read().decode())["flushed"])


if __name__ == "__main__":
    fs = OpenFilesClient()
    fs.write("/python-demo.txt", b"hello from python\n")
    print(fs.list("/"))
    print(fs.read("/python-demo.txt").decode())

File: examples/bindings/go/openfiles_client.go
package openfiles

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type Client struct { BaseURL string }

type DirEntry struct {
	Name string `json:"name"`
	Path string `json:"path"`
	Kind string `json:"kind"`
	Size uint64 `json:"size"`
}

func New(baseURL string) *Client {
	if baseURL == "" { baseURL = "http://127.0.0.1:8787" }
	return &Client{BaseURL: strings.TrimRight(baseURL, "/")}
}

func (c *Client) endpoint(prefix, path string) string {
	clean := strings.TrimLeft(path, "/")
	if clean == "" { return c.BaseURL + prefix }
	return c.BaseURL + prefix + "/" + url.PathEscape(clean)
}

func (c *Client) List(path string) ([]DirEntry, error) {
	resp, err := http.Get(c.endpoint("/v1/list", path))
	if err != nil { return nil, err }
	defer resp.Body.Close()
	if resp.StatusCode >= 300 { b,_ := io.ReadAll(resp.Body); return nil, fmt.Errorf("openfiles: %s", b) }
	var out []DirEntry
	return out, json.NewDecoder(resp.Body).Decode(&out)
}

func (c *Client) Read(path string) ([]byte, error) {
	resp, err := http.Get(c.endpoint("/v1/read", path))
	if err != nil { return nil, err }
	defer resp.Body.Close()
	if resp.StatusCode >= 300 { b,_ := io.ReadAll(resp.Body); return nil, fmt.Errorf("openfiles: %s", b) }
	return io.ReadAll(resp.Body)
}

func (c *Client) Write(path string, data []byte) error {
	req, err := http.NewRequest(http.MethodPut, c.endpoint("/v1/write", path), bytes.NewReader(data))
	if err != nil { return err }
	resp, err := http.DefaultClient.Do(req)
	if err != nil { return err }
	defer resp.Body.Close()
	if resp.StatusCode >= 300 { b,_ := io.ReadAll(resp.Body); return fmt.Errorf("openfiles: %s", b) }
	return nil
}

func (c *Client) Flush() error {
	resp, err := http.Post(c.BaseURL+"/v1/flush", "application/json", nil)
	if err != nil { return err }
	defer resp.Body.Close()
	if resp.StatusCode >= 300 { b,_ := io.ReadAll(resp.Body); return fmt.Errorf("openfiles: %s", b) }
	return nil
}
